### PR TITLE
Check for journal errors in caasp

### PR DESCRIPTION
**Pull request by thehejik**
New test for CASP journal_check.pm is trying to find suspicious lines in "journalclt -p err" (Error loglevel) output and then reports them in openQA webUI by custom write_detail_output() function.

Poo#16176

Test output available at http://dhcp209.suse.cz/tests/1426

### CODE REVIEW

File: tests/casp/journal_check.pm [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# SUSE's openQA tests
+#
+# Copyright © 2016 SUSE LLC
```
Comment (kravciak) [inline tests/casp/journal_check.pm]: 2017 :)

**Comment by kravciak** [comment]
LGTM

File: tests/casp/journal_check.pm [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    push @{$self->{details}}, {result => 'unk'};    # Is it needed?
+    push @{$self->{details}}, $detail;
+
+    open my $fh, '>', bmwqemu::result_dir() . "/$filename";
+    print $fh $output;
+    close $fh;
+
+    # Set overall result for the test, values are "ok" and "fail", result in details table has value "softfail" in addition
+    # when result in details is "ok" then overall result is "undef" and still the result in webUI is calculated as "passed"
+    $self->{result} = "fail" if $result eq "fail";
```
Comment (thehejik) [inline tests/casp/journal_check.pm]: Still not sure about this (mark overall result as failed but only if one of the "detail" is "fail" - seems it works correctly for important=1 tests).

File: tests/casp/journal_check.pm [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    my $journal_output = script_output("journalctl --no-pager -p err -b 0 | tail -n +2\n");
+
+    # Find lines which matches to the pattern_bug
+    while (my ($bug, $pattern) = each %$bug_pattern) {
+        my $buffer = "";
+        foreach my $line (split(/\n/, $journal_output)) {
+            $buffer .= $line . "\n" if ($line =~ /$pattern/);
+        }
+        $self->write_detail_output($bug, $buffer, "softfail") if $buffer;
```
Comment (kravciak) [inline tests/casp/journal_check.pm]: Maybe you can use record_soft_failure "$bug \n $buffer"

Comment (thehejik) [inline tests/casp/journal_check.pm]: yes, but for softfail issues only, I would like to use the function for ok and fail results as well.

File: tests/casp/journal_check.pm [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+# without any warranty.
+
+# Summary: Basic CASP journal tests
+# Maintainer: Tomas Hehejik <thehejik@suse.com>
+
+use base "opensusebasetest";
+use strict;
+use testapi;
+
+sub write_detail_output {
```
Comment (okurz) [inline tests/casp/journal_check.pm]: sorry I don't understand at all what is the purpose of this method

Comment (thehejik) [inline tests/casp/journal_check.pm]: it's a variant of record_soft_failure from testapi -writes an output into text box in webui with any title and result.

Comment (okurz) [inline tests/casp/journal_check.pm]: I see. good idea. It is also a good idea to try it out first in special casp test, then move it "upwards", e.g. first into something like `lib/utils.pm` as part of the tests distribution and then further to the test api of os-autoinst. I think this is what you intended :-)

**Comment by thehejik** [comment]
I change a bit overall result logic (thanks @nadvornik) and add one input check for the result in write_detail_output function.

File: tests/casp/journal_check.pm (hunk 4 above)
Comment (thehejik) [inline tests/casp/journal_check.pm]: Yes, it was meant so. But I would like to use the function first for casp only, later we can move upwards.

File: tests/casp/journal_check.pm [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    # Find lines which doesn't match to the pattern_bug by using master_pattern
+    foreach my $line (split(/\n/, $journal_output)) {
+        $self->write_detail_output("Unknown issue", $line, "fail") if ($line !~ /$master_pattern/);
+    }
+
+    # Write full journal output for reference and upload it into Uploaded Logs section in test webUI
+    script_run("journalctl --no-pager -b 0 > /tmp/full_journal.log");
+    upload_logs "/tmp/full_journal.log";
+
+    # Check for failed systemd services and examine them
```
Comment (thehejik) [inline tests/casp/journal_check.pm]: Bellow is a new stuff for checking failed systemd services, see http://dhcp209.suse.cz/tests/1702 for sample output

**Comment by okurz** [comment]
looks pretty cool, let's see how it behaves in production :-)

File: tests/casp/journal_check.pm (hunk 4 above)
Comment (okurz) [inline tests/casp/journal_check.pm]: something like https://github.com/os-autoinst/os-autoinst/pull/716 ?

Comment (thehejik) [inline tests/casp/journal_check.pm]: Yeah, something like this. But I don't know the consequences so please ask to someone else, I'm just openqa/perl newbie :-)

**Pull request merged**